pkg/project/auth/watch.go

```diff
@@ -0,0 +1,196 @@
+
+func NewUserProjectWatcher(username string, groups []string, projectCache *projectcache.ProjectCache, authCache *AuthorizationCache) *userProjectWatcher {
+	userInfo := &user.DefaultInfo{Name: username, Groups: groups}
+	namespaces, _ := authCache.List(userInfo)
+	namespaceNames := sets.String{}
+	for _, namespace := range namespaces.Items {
+		namespaceNames.Insert(namespace.Name)
+	}
+
+	w := &userProjectWatcher{
+		username: username,
+		groups:   groups,
+
+		cacheIncoming: make(chan watch.Event, 100),
+		cacheError:    make(chan error, 1),
+		outgoing:      make(chan watch.Event),
+		userStop:      make(chan struct{}),
+
+		projectCache:  projectCache,
+		authCache:     authCache,
+		knownProjects: namespaceNames,
+	}
+	w.emit = func(e watch.Event) {
+		w.outgoing <- e
+	}
+	return w
+}
+
+func (w *userProjectWatcher) GroupMembershipChanged(namespaceName string, removedUsers, removedGroups, addedUsers, addedGroups sets.String) {
+	removed := removedUsers.Has(w.username) || removedGroups.HasAny(w.groups...)
+	added := addedUsers.Has(w.username) || addedGroups.HasAny(w.groups...)
+
+	switch {
+	// do nothing if we're net neutral
+	case removed && added:
+	case !removed && !added:
+
+	case removed:
+		if !w.knownProjects.Has(namespaceName) {
+			return
+		}
+		w.knownProjects.Delete(namespaceName)
+		select {
+		case w.cacheIncoming <- watch.Event{
+			Type:   watch.Deleted,
+			Object: convertNamespace(&kapi.Namespace{ObjectMeta: kapi.ObjectMeta{Name: namespaceName}}),
+		}:
+
+		case <-time.After(10 * time.Millisecond):
+			w.cacheError <- errors.New("delete notification timeout")
+		}
+
+	case added:
+		if w.knownProjects.Has(namespaceName) {
+			return
+		}
+		w.knownProjects.Insert(namespaceName)
+		namespace, err := w.projectCache.GetNamespace(namespaceName)
+		if err != nil {
+			util.HandleError(err)
+			return
+		}
+
+		select {
+		case w.cacheIncoming <- watch.Event{
+			Type:   watch.Added,
+			Object: convertNamespace(namespace),
+		}:
+
+		case <-time.After(10 * time.Millisecond):
+			w.cacheError <- errors.New("add notification timeout")
+		}
+
+	}
+
+}
+
+func convertNamespace(namespace *kapi.Namespace) *projectapi.Project {
+	return &projectapi.Project{
+		ObjectMeta: namespace.ObjectMeta,
+		Spec: projectapi.ProjectSpec{
+			Finalizers: namespace.Spec.Finalizers,
+		},
+		Status: projectapi.ProjectStatus{
+			Phase: namespace.Status.Phase,
+		},
+	}
+}
+
+// translate pulls stuff from etcd, converts, and pushes out the outgoing channel. Meant to be
+// called as a goroutine.
+func (w *userProjectWatcher) Watch() {
+	defer close(w.outgoing)
+	defer util.HandleCrash()
+
+	for {
+		select {
+		case err := <-w.cacheError:
+			if err != nil {
+				w.emit(watch.Event{
+					Type: watch.Error,
+					Object: &unversioned.Status{
+						Status:  unversioned.StatusFailure,
+						Message: err.Error(),
+					},
+				})
+			}
+			return
+
+		case <-w.userStop:
+			w.authCache.RemoveWatcher(w)
```

Comment:
> So in a panic case, the watcher won't be removed?

Will move to `defer` if the UI team has a desire for a half working watch on projects.  This was just so they'd have something to actually see what they code with it and decide if it had value before I invested the time in filling it out.

Making a coherent watch that also tracks changes to `namespaces` and funnels them through is possible, but the simplest implementation requires introspection of `resourceVersion` and an understanding of its ordered nature.  If they want that, am I cleared to parse the string?